design/ceph-osd-redesign-for-octopus.md

```diff
@@ -0,0 +1,342 @@
+to the Ceph Cluster CRD only require changes to the CRD and to the pods created by the operator,
+cutting the number of locations where code is modified in half.
+
+#### Stage 4: Host device OSD inspection
+Rook can determine which OSDs are prepared on the host by running `ceph-volume lvm list` in the same
+manner as **Stage 1**. The result is similarly stored in a ConfigMap may now report existing OSDs on
+the host. These OSDs may have been created by Rook, or they may have been created manually or via an
+action of Ceph's Rook orchestrator.
+
+#### Stage 5: Start OSDs
+The last step remaining is to start the OSDs which were found by `ceph-volume`'s prior inspection.
+The device inspection report stored in **Stage 4** will be used to determine which OSDs on the host
+belong to the operator's Ceph cluster (by UUID examination), and appropriate deployments will be
+created for each OSD. The report should contain the information necessary to determine which flags
+should be passed to `ceph-osd`.
+
+If possible, the existing trend of using init containers instead of a separate Rook binary
+(trend established for Ceph mons, mgrs, MDSes, and RGWs) should be used here with the intent being
+again to keep most of the logic in the operator. If teardown operations are needed, the Kubernetes
+container `PreStop` lifecycle hook should be used if reasonably possible.
+
+#### Shared code
+**Stage 1** and **Stage 4** have the same function: run a command and put the result in a ConfigMap.
+Implementing this as a building block would allow for reuse in other parts of Rook and in the
+future. This could be used immediately for `ceph version` reporting at the beginning of
+orchestration without relying on the pod logs being intact, as has been a problem in some user
+environments.
+
+
+Deprecation plan
+--------------------
+The first step in the deprecation plan is to have the new design in-place. Once the new design is
+available, the new design should always be used for any new OSD devices. Legacy OSDs will
+not be able to be created, but existing legacy OSDs will still be run using the legacy code path.
+
+Users will be informed in the Rook v1.1 release notes of the intention to deprecate directory-based
+OSDs and Filestore as a backing store and informed of the progress toward that goal. Before support
+is totally removed for the deprecated features, users should have OSD migration documentation
+available, and they should have had a good amount of time to migrate their OSDs.
+
+Once the new OSD design is operational in Rook and is the only code path used for new
+OSDs, the deprecation will be official, and that should be reported to users in the release notes
+and upgrade guide. After there is documentation for manually migrating OSDs, we can expect that the
+following version of Rook will officially drop support for legacy OSDs.
+
+It is ideal if users have access to the migration utility at the time of deprecation; however, the
```

Comment:
Again, if we can detect and activate legacy OSDs, there is no migration to do and no tool to write.
Only preparing using the old fashion way will not be possible anymore.

Reply:
The document was revised to only deprecate directory and filestore disks. Not preexisting bluestore disks.